### ISSUE TITLE
模型大规模训练的时候，内存也会一直增加

### ISSUE THREAD

**Issue by 404289680** [JDAI-CV/fast-reid#217]
10wid直接超时中断了，从十万中抽3万id可以训练

**Comment by L1aoXingyu** [comment]
Yes, this is a problem that I just found.

The main problem is below 
https://github.com/JDAI-CV/fast-reid/blob/3549685c4c25e2ff41c4e49696b034f53c0ab14a/fastreid/data/samplers/triplet_sampler.py#L140

It will prepare the whole complete epoch index then begin training. So when the dataset is very large, it needs a long time to prepare.

I am trying to fix this problem, if you have any suggestion, you can discuss it with me here.

**Comment by 404289680** [comment]
能不能把prepare独立出来，保存成一个文件，到时候读取这个文件

**Comment by L1aoXingyu** [comment]
目前有三个地方的数据处理会耗时

第一个是下面这里的 relabel，这个功能主要是为了 multi-dataset 训练的时候可以把多个 domain 的 label 重新排列一下，如果你可以出来数据集的统计信息，那么下面这个部分就通过了，如果需要很长的时间做 relabel，可以考虑自己先把 label 处理好得到一个 txt，然后传进去，不用做 relabel。

https://github.com/JDAI-CV/fast-reid/blob/9c667d1a0fbd8a6b5a82ad4d04d51b47d6b2957a/fastreid/data/common.py#L22

第二个耗时的地方是 sampler 里面的初始化，需要对每个 id 建立对应 index 的字典

https://github.com/JDAI-CV/fast-reid/blob/9c667d1a0fbd8a6b5a82ad4d04d51b47d6b2957a/fastreid/data/samplers/triplet_sampler.py#L111

第三个耗时的地方是 sampler 里面每个 epoch 完成之后会生成下一个 epoch 完整的 index，这个地方不能独立出来，因为每次 epoch 不仅需要 shuffle，还需要重新采样每个 id 下面的 index，目前的做法是每次只准备当前 iter 的 index，具体可以看看下面的 code

https://github.com/JDAI-CV/fast-reid/blob/9c667d1a0fbd8a6b5a82ad4d04d51b47d6b2957a/fastreid/data/samplers/triplet_sampler.py#L166

**Comment by 404289680** [comment]
感谢，但是这代码还有个小bug，fast-reid/fastreid/data/common.py，这个文件里27行if item[1] in pids这个时间复杂度过高了，建议加一个hash set，改成if item[1] in set

**Comment by L1aoXingyu** [comment]
以前的版本用的 set，但是用 set 的时候，里面的 id 顺序会乱掉，如果训练中断之后 resume training，id 会全部对不上，所以以前存了一个 pid2label 的 pickle 文件，现在使用 list 可以避免额外存一个文件。这里我再思考一下如何处理。

你现在可以正常训练 10w ID 吗？

**Comment by 404289680** [comment]
不用set无法训练，用了set可以训练

**Comment by 404289680** [comment]
sets = set()
        if self.relabel:
            pids = list()
            for i, item in enumerate(img_items):
                if item[1] in sets: continue
                pids.append(item[1])
                sets.add(item[1])
            self.pids = pids

**Comment by L1aoXingyu** [comment]
这样可以，一个 hash table 一个 list

Good point!

**Comment by 404289680** [comment]
这里的耗时用了hash后很短，不用就会卡这了，我有200多万图片，o(n^2),你可以估计下耗时

**Comment by L1aoXingyu** [comment]
Thanks, I see!

**Comment by L1aoXingyu** [comment]
sampler 这里如果你不修改的话，应该会出现周期性卡顿。

**Comment by 404289680** [comment]
10w ID  8卡要训一个星期正常吗,且打印出来的time从前期的0.14到现在的0.28在增加，显存也在增加

**Comment by L1aoXingyu** [comment]
> 10w ID  8卡要训一个星期正常吗,且打印出来的time从前期的0.14到现在的0.28在增加，显存也在增加

训练时间正常的，但是显存是不应该增加的

**Comment by 404289680** [comment]
显存从刚开始的时候10.1G ， 过了段时间变成了10.9G，后来稳定了，但是时间一直在增加，请问每次迭代的时间一直在增加的原因是什么

**Comment by 404289680** [comment]
提个小建议，弄个特征维度的参数，不然2048维度，id一多很恐怖

**Comment by L1aoXingyu** [comment]
> 提个小建议，弄个特征维度的参数，不然2048维度，id一多很恐怖

有一个 `reductionHead` 可以做降维

> 显存从刚开始的时候10.1G ， 过了段时间变成了10.9G，后来稳定了，但是时间一直在增加，请问每次迭代的时间一直在增加的原因是什么

每次迭代时间增加我也没有找到原因，我还需要继续 check 一下

**Comment by 404289680** [comment]
找到这个reductionHead了，谢谢

**Comment by 404289680** [comment]
reductionHead改了没生效。。

**Comment by L1aoXingyu** [comment]
把 `BNNeckHead` 改成 `ReductionHead` 之后，改一下 `REDUCTION_DIM` 就可以了。

**Comment by 404289680** [comment]
好的，那个训练迭代时间越来越长，应该是某个地方gpu计算越来越多的原因，具体是哪个地方，我还得再看看

**Comment by 404289680** [comment]
我说错了，迭代时间是忽快忽慢，我怀疑是sampler的问题，我再看看

**Comment by L1aoXingyu** [comment]
迭代时间是一个滑动平均的更新方式，所以并不是完全准确地反映当前 iter 的时间。

另外，如果 `sampler` 里面使用 `NAIVE_WAY`，那么下面这个地方会出现不定时的延时

https://github.com/JDAI-CV/fast-reid/blob/fd06a2819a1973b54cd86e7332780660ab09f0fb/fastreid/data/samplers/triplet_sampler.py#L153

**Comment by 404289680** [comment]
我看了这段代码，感觉耗时应该是正常的，我的怀疑应该是错的，我找找其他地方

**Comment by 404289680** [comment]
if len(avai_idxs) < self.num_instances: avai_pids.remove(pid)  这行代码有可能会耗时比较多，因为list.remove是o(n)级别的

**Comment by L1aoXingyu** [comment]
Maybe `set` is a good choice.

**Comment by L1aoXingyu** [comment]
Maybe `set` is an another choice.

**Comment by 404289680** [comment]
是的，我试了set，时间现在比较稳定了，但是我得再跑久一点才能得出结论

**Comment by 404289680** [comment]
是因为这个，现在没问题了

**Comment by L1aoXingyu** [comment]
我又做了进一步的修改如下

https://github.com/JDAI-CV/fast-reid/blob/07872d4cdb5532395e7c67fdf15477987fcc194d/fastreid/data/common.py#L20

因为我发现不同的机器上跑的程序生成的 `pid_dict` 是不同的，所以我多做了一个 `sort`。

**Comment by L1aoXingyu** [comment]
我发现 code 里面有 memory leak，RAM 一直在涨，具体是哪里的问题我还没发现。

**Comment by 404289680** [comment]
我本来以为没问题了，当时单卡跑了一晚上确实是正常的，我八卡跑了一晚上时间一直是在涨的，奇怪

**Comment by L1aoXingyu** [comment]
我可能需要进一步排查一下，看看单卡是不是没有 memory leak 的问题

**Comment by 404289680** [comment]
np.random.choice(list(avai_pids), self.num_pids_per_batch, replace=False)
这行代码，如果不做list(avai_pids)这个操作，会报错，做了这个操作，耗时会比较多，你有没有什么好的办法

**Comment by L1aoXingyu** [comment]
这里加上 `list` 之后，多卡训练 id 会乱掉，因为 `set` 本身是无序的，所以 `list` 每个 process 里面的顺序都不同，random choice 出来的 id 都是不同的。这里我暂时还没有想到怎么搞， 现在还是先用 list 做一下。

目前我比较头疼 memory 一直增加的问题。

**Comment by 404289680** [comment]
memory 一直增加其实还好，我上周那个，跑了一个星期了，memory也没多多少出来

**Comment by L1aoXingyu** [comment]
你有监控吗？为啥我这边跑两天 memory 到上 100G 了。

**Comment by 404289680** [comment]
我最新代码10wid刚跑就花了140G了，上周那代码，我跑了一周，现在是150G，没啥区别

**Comment by L1aoXingyu** [comment]
我现在准备试试 random sample，看看时间会不会一直增加。

**Comment by 404289680** [comment]
我把np.random.choice(list(avai_pids), self.num_pids_per_batch, replace=False)这个
改成list(avai_pids)[:self.num_pids_per_batc]
耗时感觉就正常了，但是这个就是做个验证

**Comment by ivylinden** [comment]
直接读一个行人ID的txt文件比较快吧，把多个数据集的标签生成为一个txt，txt里存放图片路径 ID 提及cam信息

**Comment by L1aoXingyu** [comment]
> 直接读一个行人ID的txt文件比较快吧，把多个数据集的标签生成为一个txt，txt里存放图片路径 ID 提及cam信息

直接读 txt，每次 sampler 还是需要做

**Comment by HouBiaoLiu** [comment]
可不可以用INPLACE-ABN来代替这里的ibn模块 减少显存消耗

**Comment by L1aoXingyu** [comment]
> 可不可以用INPLACE-ABN来代替这里的ibn模块 减少显存消耗

其实并不是显存的问题，而是内存的问题。另外 inplace_abn 有一些坑，我之前踩过，所以不太想迁移过来。

**Comment by HouBiaoLiu** [comment]
请问我看您之前做aicity比赛时，提到了弱监督车辆检测来做数据增强，在这个工程里怎么没用到呢，用了之后rank1,mAP指标会不会再提升几个点 @L1aoXingyu

**Comment by 404289680** [comment]
请问time是在哪统计的,是指model infer的时间还是总时间？

**Comment by 404289680** [comment]
我测了下耗时，增加的部分是model infer过程的耗时，但是具体是model的哪个模块，backbone还是head，还不清楚

**Comment by 404289680** [comment]
我知道问题在哪了，这个不算是bug，时间本来就会增加，因为你这有个FREEZE_ITERS: 10，freeze结束后耗时自然要增加，哎

**Comment by L1aoXingyu** [comment]
> 我知道问题在哪了，这个不算是bug，时间本来就会增加，因为你这有个FREEZE_ITERS: 10，freeze结束后耗时自然要增加，哎，幸幸苦苦找那么久，白搞，现在的问题就剩那个set不知道如何优雅的解决

我之前试过把 freeze_iter 调成 0，确实没有出现时间增加的问题。

**Comment by L1aoXingyu** [comment]
> 请问我看您之前做aicity比赛时，提到了弱监督车辆检测来做数据增强，在这个工程里怎么没用到呢，用了之后rank1,mAP指标会不会再提升几个点

我没有参加过 aicity 😂

**Comment by 404289680** [comment]
请问有没有支持fp16训练的计划

**Comment by L1aoXingyu** [comment]
> 请问有没有支持fp16训练的计划

正在支持，目前有一些 code 在重构，下周应该就会把 fp16 集成上去。

**Comment by L1aoXingyu** [comment]
已经支持 amp fp16 training @404289680

**Comment by L1aoXingyu** [comment]
It seems CPU memory gradually leaks due to this issue.

https://github.com/pytorch/pytorch/issues/13246

**Comment by TCHeish** [comment]
using  AMP training this occured at first training step
------------------------------------------------------------------------
/usr/local/lib/python3.6/dist-packages/torch/optim/lr_scheduler.py:123: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)

------------------------------------------------------------------------
according to this, first step optimizer is skipped, but lr_scheduler still update
https://discuss.pytorch.org/t/optimizer-step-before-lr-scheduler-step-error-using-gradscaler/92930

do you have better solution for this ?

**Comment by L1aoXingyu** [comment]
@TCHeish I think skipping the first optimizer step would have no negative effect, and it's normal when using amp training. Because the first few steps will determine the proper loss scaler.

**Comment by TCHeish** [comment]
Hi, 

Is that work well in production after training 10wid datasets?
(For example, the rank1 & mAp compared to 1wid model)

Thanks

**Comment by L1aoXingyu** [comment]
> Is that work well in production after training 10wid datasets?
> (For example, the rank1 & mAp compared to 1wid model)

I conduct an experiments on 3w IDs, and amp training results are the same as the FP32 training.

**Comment by 404289680** [comment]
AMP_ENABLED=True
我设置amp后为啥显存用的跟多，速度也更慢了，你的可以省显存么

**Comment by L1aoXingyu** [comment]
可以先确认一下 GPU 架构是否支持半精度训练，比如 tesla P40 就不支持

**Comment by 404289680** [comment]
2080ti不可以么，似乎还跟模型相关，resnet可以省显存，resnest显存更多了

**Comment by L1aoXingyu** [comment]
2080ti 支持 fp16 的，应该比较通用的结构支持更好，resnet 训练速度能提升吗？

**Comment by 404289680** [comment]
resnet可以，resnest显存用的更多了

**Comment by 404289680** [comment]
单卡结果和多卡结果也不一样，多卡省了显存，但是速度更慢了

**Comment by 404289680** [comment]
https://zhuanlan.zhihu.com/p/165152789，看评论似乎跟我一样的，好像还是走apex靠谱一些

**Comment by L1aoXingyu** [comment]
那我还是用 apex 替换 pytorch 官方的 amp 吧。

**Comment by TCHeish** [comment]
> 那我还是用 apex 替换 pytorch 官方的 amp 吧。

請問用apex需要更改那些地方?(r50-ibn的話)

**Comment by L1aoXingyu** [comment]
你可以参考 nvidia apex 官方文档
https://github.com/NVIDIA/apex
